### PR TITLE
Project 09

### PR DESCRIPTION
**Pull request by elilutsky**

### CODE REVIEW

File: projects/09/Snake/Block.jack [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+    /** Deallocates the object's memory. */
+    method void dispose() {
+        do Memory.deAlloc(this);
+        return;
+    }
+
+    /** Draws the square on the screen. */
+    method void draw() {
+		let isVisible = true;
+        do Screen.setColor(true);
```
Comment (denisgolyanov) [inline projects/09/Snake/Block.jack]: you can extract these to a common function which uses the ```isVisible``` field

File: projects/09/Snake/BlockList.jack [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+		if (next = null) {
+            let next = blockIn;
+			return;
+        }
+		do next.appendBlockList(blockIn);
+		return;
+	}
+	
+	/** Removes the last element from the list and erases the block. This
+		method expects the list to contain more that 1 element. */
```
Comment (denisgolyanov) [inline projects/09/Snake/BlockList.jack:50]: document return value

File: projects/09/Snake/Game.jack [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+		do snake.dispose();
+		do rand.dispose();
+		do food.dispose();
+        do Memory.deAlloc(this);
+        return;
+    }
+	
+	method void createSnake() {
+		let snake = Snake.new(15);
+		do snake.initBasicSnake(200, 120);
```
Comment (denisgolyanov) [inline projects/09/Snake/Game.jack]: why isn't initBasicSnake part of the CTOR?

File: projects/09/Snake/Game.jack [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+    method void dispose() {
+		do snake.dispose();
+		do rand.dispose();
+		do food.dispose();
+        do Memory.deAlloc(this);
+        return;
+    }
+	
+	method void createSnake() {
+		let snake = Snake.new(15);
```
Comment (denisgolyanov) [inline projects/09/Snake/Game.jack]: you can define the constant as self describing static fields

File: projects/09/Snake/Game.jack [hunk 5]
```diff
@@ -0,0 +1,153 @@
...
+		do snake.initBasicSnake(200, 120);
+		return;
+	}
+	
+	method void createFood() {
```
Comment (denisgolyanov) [inline projects/09/Snake/Game.jack:34]: document the method, what does radius mean in terms of food?

File: projects/09/Snake/Game.jack [hunk 6]
```diff
@@ -0,0 +1,153 @@
...
+		let y = rand.randRange(bottomLimit - topLimit - radius) + topLimit + radius;
+		let food = Food.new(x, y, radius);
+		let foodExists = true;
+		do food.draw();
+		return;
+	}
+	
+	/** Returns true if the snake collides with the food */
+	method boolean detectFoodCollison() {
+		if (snake.doesCollide(food.getY() - food.getRadius() - 5, food.getY() + food.getRadius() + 5, food.getX() - food.getRadius() - 5, food.getX() + food.getRadius() + 5)) { return true; }
```
Comment (denisgolyanov) [inline projects/09/Snake/Game.jack]: what does the 5 mean?
+ break into multiple lines

File: projects/09/Snake/Snake.jack [hunk 7]
```diff
@@ -0,0 +1,161 @@
...
+    method void dispose() {
+		do blockList.dispose();
+        do Memory.deAlloc(this);
+        return;
+    }
+	
+	method void initBasicSnake(int startLocationX, int startLocationY) {
+		var Block first, second, third;
+		let first = Block.new(startLocationX, startLocationY, blockSize);
+		let second = Block.new(startLocationX + blockSize + 5, startLocationY, blockSize);
```
Comment (denisgolyanov) [inline projects/09/Snake/Snake.jack]: what does the 5 mean? why is init seperated from CTOR ?

Comment (denisgolyanov) [reply projects/09/Snake/Snake.jack]: maybe extract the 5 to a static const which will be self describing

File: projects/09/Snake/Snake.jack [hunk 8]
```diff
@@ -0,0 +1,161 @@
...
+	
+	/** Returns true if the snake's head collides with the given area */
+	method boolean doesCollide(int top, int bottom, in left, int right) {
+		var int snakeTop, snakeLeft, snakeRight, snakeBottom;
+		let snakeTop = getHeadTop();
+		let snakeLeft = getHeadLeft();
+		let snakeRight = getHeadRight();
+		let snakeBottom = getHeadBottom();
+		
+		if ((snakeTop > top) & (snakeTop < bottom)) {
```
Comment (denisgolyanov) [inline projects/09/Snake/Snake.jack:132]: how can the head of the snake be both above the top limit and below the bottom limit at the same time?
am I missing something?

regardless, document the meaning of each if condition

**Review by denisgolyanov — CHANGES_REQUESTED** [review]

File: projects/09/Snake/Game.jack (hunk 4 above)
Comment (elilutsky) [reply projects/09/Snake/Game.jack]: I solved it by extracting the block size and starting position to be CTOR arguments

File: projects/09/Snake/Game.jack (hunk 6 above)
Comment (elilutsky) [reply projects/09/Snake/Game.jack]: Simple padding that makes the game easier to play (The player will get the food even if not exactly on it).

File: projects/09/Snake/Snake.jack (hunk 8 above)
Comment (elilutsky) [reply projects/09/Snake/Snake.jack:132]: The top limit is actually lower than the bottom limit (The point [0,0] is located at the top left corner of the screen)

**Pull request merged**